### PR TITLE
feat: add edit-code and view-code component

### PR DESCRIPTION
**Pull request by luckyman20**
## Description
Added view-code and edit-code component, designed their screens and added functionality too.
Functionalities added are -

- View Code
- Edit Code
- Delete Code
- View Code Value
- Edit Code Value
- Delete Code Value
 
## Related issues and discussion
#379

## Screenshots, if any

view code
_View Code_

edit code
_Edit Code_

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

File: src/app/system/codes/view-code/view-code.component.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+   * @param {number} index Index of the row.
+   */
+  disableRow(index: number) {
+    this.codevalues.at(index).disable();
+    this.codeValueRowStatus[index] = 'disabled';
+    this.codevalues.at(index).markAsPristine();
+  }
+
+  /**
+   * Adds/Initalises the given code value.
```
Comment (abhaychawla) [inline src/app/system/codes/view-code/view-code.component.ts]:
```suggestion
   * Adds the given code value.
```

File: src/app/system/codes/view-code/view-code.component.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+      });
+  }
+
+  /**
+   * Enables the given row.
+   * @param {number} index Index of the row.
+   */
+  enableRow(index: number) {
+    this.codevalues.at(index).enable();
+    this.codeValueRowStatus[index] = 'editing';
```
Comment (abhaychawla) [inline src/app/system/codes/view-code/view-code.component.ts]:
```suggestion
    this.codeValueRowStatus[index] = 'edit';
```

File: src/app/system/system-routing.module.ts [hunk 3]
```diff
@@ -34,27 +38,49 @@ const routes: Routes = [
       path: 'system',
       data: { title: extract('System'), breadcrumb: 'System' },
       children: [
-      {
```
Comment (abhaychawla) [inline src/app/system/system-routing.module.ts]: Remove unnecessary changes.

File: src/app/system/codes/view-code/view-code.component.html [hunk 4]
```diff
@@ -0,0 +1,87 @@
+<div class="container m-b-20" fxLayout="row" fxLayoutAlign="end" fxLayoutGap="20px">
+
+  <button mat-raised-button color="primary" (click)="addCodeValuesRow()">
```
[the remark this answers is not in the document]
Comment (luckyman20) [reply src/app/system/codes/view-code/view-code.component.html]: There is already another function with the same name.

File: src/app/system/codes/view-code/view-code.component.html [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                Name is <strong>required</strong>
+              </mat-error>
+            </mat-form-field>
+
+            <div class="active-wrapper" fxFlex="8%">
+              <mat-checkbox formControlName="active" class="active">Active</mat-checkbox>
+            </div>
+
+            <div fxLayoutGap="2%" fxFlex="22%" class="options-wrapper">
+              <button mat-stroked-button color="primary" *ngIf="codeValueRowStatus[i] === 'editing'" [disabled]="codevalues.at(i).pristine" (click)="updateCodeValue(i)">
```
[the remark this answers is not in the document]
Comment (luckyman20) [reply src/app/system/codes/view-code/view-code.component.html]: A stroked button is looking a lot better than a raised button in this case.

File: src/app/system/codes/view-code/view-code.component.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+  /**
+   * Removes/Deletes the particular code value.
+   * @param {number} index Index of the row.
+   */
+  removeCodeValue(index: number) {
+    const codeValueId = this.codeValuesData[index].id;
+    this.systemService.deleteCodeValue(this.codeData.id, codeValueId)
+      .subscribe((response: any) => { });
+    this.codeValuesData.splice(index, 1);
```
[the remark this answers is not in the document]
Comment (luckyman20) [reply src/app/system/codes/view-code/view-code.component.ts]: Asked the same question on the last PR.

File: src/app/system/codes/view-code/view-code.component.html [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                Submit
+              </button>
+              <button mat-button *ngIf="codeValueRowStatus[i] === 'editing'" (click)="disableRow(i)">
```
[the remark this answers is not in the document]
Comment (luckyman20) [reply src/app/system/codes/view-code/view-code.component.html]: This looks better than a raised button here.

File: src/app/system/system-routing.module.ts (hunk 3 above)
Comment (luckyman20) [reply src/app/system/system-routing.module.ts]: There are no unnecessary changes in this file as of now, multiple lines are being removed because it was the first component.

File: src/app/system/codes/view-code/view-code.component.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (abhaychawla) [reply src/app/system/codes/view-code/view-code.component.html]: Alright.

File: src/app/system/codes/view-code/view-code.component.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (abhaychawla) [reply src/app/system/codes/view-code/view-code.component.ts]: Use `map` and `pipe` operators.

File: src/app/system/system-routing.module.ts (hunk 3 above)
Comment (abhaychawla) [reply src/app/system/system-routing.module.ts]: There are changes to indentation. Refactoring the code is not a part of this issue.

File: src/app/system/codes/view-code/view-code.component.html [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+  <mat-card>
+
+    <mat-card-title>{{ codeData.name }}</mat-card-title>
+
+    <form [formGroup]="codeValuesForm">
+
+      <mat-card-content>
+
+        <div formArrayName="codevalues" *ngFor="let codevalue of codevalues.controls; let i = index;">
```
Comment (abhaychawla) [inline src/app/system/codes/view-code/view-code.component.html]:
```suggestion
        <div formArrayName="codeValues" *ngFor="let codeValue of codeValues.controls; let i = index;">
```

File: src/app/system/codes/view-code/view-code.component.html [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            </div>
+
+            <div fxLayoutGap="2%" fxFlex="22%" class="options-wrapper">
+              <button mat-stroked-button color="primary" *ngIf="codeValueRowStatus[i] === 'edit'" [disabled]="codevalues.at(i).pristine" (click)="updateCodeValue(i)">
+                Submit
+              </button>
+              <button mat-button *ngIf="codeValueRowStatus[i] === 'edit'" (click)="disableRow(i)">
+                Cancel
+              </button>
+              <button mat-icon-button color="primary" *ngIf="codeValueRowStatus[i] === 'new'" matTooltip="Add" matTooltipPosition="above" (click)="addCodeValue(i)">
```
Comment (abhaychawla) [inline src/app/system/codes/view-code/view-code.component.html]: Disable the button till mandatory fields are filled.

File: src/app/system/codes/view-code/view-code.component.scss [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+        padding: 0 0 17.5px 0;
+        position: absolute;
+        bottom: 0;
+        right: 0;
+      }
+  }
+
+}
+
+  .options-wrapper {
```
Comment (abhaychawla) [inline src/app/system/codes/view-code/view-code.component.scss]:
```suggestion
.options-wrapper {
```
Remove any extra space.

File: src/app/system/codes/view-code/view-code.component.ts [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+/** Custom Components */
+import { DeleteDialogComponent } from 'app/shared/delete-dialog/delete-dialog.component';
+
+@Component({
+  selector: 'mifosx-view-code',
+  templateUrl: './view-code.component.html',
+  styleUrls: ['./view-code.component.scss']
+})
+
+/**
```
Comment (abhaychawla) [inline src/app/system/codes/view-code/view-code.component.ts]: Incorrectly placed comment.

File: src/app/system/codes/view-code/view-code.component.ts [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+   */
+  addCodeValueRow() {
+    this.codevalues.push(this.createCodeValuesRow());
+    this.codevalues.at(this.codevalues.length - 1).enable();
+    this.codeValueRowStatus.push('new');
+  }
+
+  /**
+   * Creates a code value row in code values form.
+   * @param {string} codeValueName Code value name.
```
Comment (abhaychawla) [inline src/app/system/codes/view-code/view-code.component.ts]: Don't express the comment like this. It is misleading as it seems we need to pass 4 parameters.

File: src/app/system/codes/view-code/view-code.component.ts [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    this.codeValueRowStatus.splice(index, 1);
+  }
+
+  /**
+   * Updates the particular code value.
+   * @param {number} index Index of the row.
+   */
+  updateCodeValue(index: number) {
+    const updatedCodeValue: { name: string, description: string, position: number, isActive: boolean } = {
+      name: this.codevalues.at(index).get('name').value,
```
Comment (abhaychawla) [inline src/app/system/codes/view-code/view-code.component.ts]: Try to access this object from the array directly instead of constructing it here again.

File: src/app/system/codes/view-code/view-code.component.ts [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+    this.codevalues.at(index).markAsPristine();
+  }
+
+  /**
+   * Adds the given code value.
+   * @param {number} index Index of the row.
+   */
+  addCodeValue(index: number) {
+    const newCodeValue: { name: string, description: string, position: string, isActive: boolean } = {
+      name: this.codevalues.at(index).get('name').value,
```
Comment (abhaychawla) [inline src/app/system/codes/view-code/view-code.component.ts]: Try to access this object from the array directly instead of constructing it here again.

File: src/app/system/system-routing.module.ts [hunk 15]
```diff
@@ -34,27 +38,49 @@ const routes: Routes = [
...
+            {
+              path: 'view/:id',
+              data: { title: extract('View Code'), routeParamBreadcrumb: 'id' },
+              children: [
+                {
+                  path: '',
+                  component: ViewCodeComponent,
+                  resolve: {
+                    code: CodeResolver,
+                    codevalues: CodeValuesResolver
```
Comment (abhaychawla) [inline src/app/system/system-routing.module.ts]:
```suggestion
                    codeValues: CodeValuesResolver
```

File: src/app/system/codes/view-code/view-code.component.ts (hunk 13 above)
Comment (luckyman20) [reply src/app/system/codes/view-code/view-code.component.ts]: Can you please specify which array you are pointing to?

File: src/app/system/codes/view-code/view-code.component.ts (hunk 14 above)
Comment (luckyman20) [reply src/app/system/codes/view-code/view-code.component.ts]: Refer to the comment above.

File: src/app/system/system-routing.module.ts (hunk 3 above)
Comment (luckyman20) [reply src/app/system/system-routing.module.ts]: Sorry, but I do not see any indentation related changes.

File: src/app/system/codes/view-code/view-code.component.ts (hunk 14 above)
Comment (abhaychawla) [reply src/app/system/codes/view-code/view-code.component.ts]: `codevalues`

File: src/app/system/codes/view-code/view-code.component.ts (hunk 13 above)
Comment (abhaychawla) [reply src/app/system/codes/view-code/view-code.component.ts]: `codevalues`

File: src/app/system/system.service.ts [hunk 16]
```diff
@@ -46,6 +46,67 @@ export class SystemService {
...
+   * @param codeId Code ID.
+   * @returns {Observable<any>}
+   */
+  updateCode(code: any, codeId: string): Observable<any> {
+    return this.http.put(`/codes/${codeId}`, code);
+  }
+
+  /**
+   * @param codeId Code ID.
+   * @returns {Observable<any>}
```
Comment (abhaychawla) [inline src/app/system/system.service.ts]: Make sure you mention what is being returned by `get` function.

File: src/app/system/system-routing.module.ts (hunk 3 above)
Comment (abhaychawla) [reply src/app/system/system-routing.module.ts]: There are some changes to spacing.

File: src/app/system/codes/edit-code/edit-code.component.scss [hunk 17]
```diff
@@ -0,0 +1,3 @@
+.container {
+    max-width: 37rem;
+  }
```
Comment (abhaychawla) [inline src/app/system/codes/edit-code/edit-code.component.scss]:
```suggestion
}
```

**Comment by abhaychawla** [comment]
@luckyman20 There are still a few unresolved comments in this PR. Please have a look. Also, rebase your PR when you are done.

File: src/app/system/codes/view-code/view-code.component.ts [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+   */
+  delete() {
+    const deleteCodeDialogRef = this.dialog.open(DeleteDialogComponent, {
+      data: { deleteContext: `code ${this.codeData.name}` }
+    });
+    deleteCodeDialogRef.afterClosed().subscribe((response: any) => {
+      if (response.delete) {
+        this.systemService.deleteCode(this.codeData.id)
+          .subscribe(() => {
+            this.router.navigate(['../../'], { relativeTo: this.route });
```
Comment (abhaychawla) [inline src/app/system/codes/view-code/view-code.component.ts]: The route redirect is not working. Please check.
